### PR TITLE
fix: merge .percy.yml config options with snapshot options for serializeDOM

### PR DESCRIPTION
**Pull request by rishigupta1599**
## Summary
- Config options from `.percy.yml` were not being passed to `PercyDOM.serialize()`
- Only per-snapshot options were used for DOM serialization
- Now merges `@cli_config['snapshot']` with user options, with per-snapshot options taking priority

## Test plan
- [ ] Existing tests pass
- [ ] Verify `.percy.yml` config options are applied during DOM serialization

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/percy.rb [hunk 1]
```diff
@@ -79,10 +79,15 @@ def self.snapshot(driver, name, options = {})
     begin
       percy_dom_script = fetch_percy_dom
       driver.execute_script(percy_dom_script)
-      dom_snapshot = if responsive_snapshot_capture?(options)
-        capture_responsive_dom(driver, options, percy_dom_script: percy_dom_script)
+
+      # Merge .percy.yml config options with snapshot options (snapshot options take priority)
+      config_options = @cli_config&.dig('snapshot') || {}
+      merged_options = config_options.merge(options)
```
Comment (rishigupta1599) [inline lib/percy.rb]: **[High] String/symbol key collision breaks per-call merge priority**

`@cli_config` is JSON-parsed, so `config_options` has **string** keys, while per-call `options` conventionally use **symbol** keys. `Hash#merge` treats `'widths'` and `:widths` as distinct, so `config_options.merge(options)` keeps both rather than letting the per-call value override the YAML one. Downstream consumers read symbol keys only (e.g. `options[:widths]`, `responsive_snapshot_capture?`), so the YAML config is never seen through the merge and the documented per-call priority does not actually hold. This defeats the core purpose of the PR.

**Suggestion:** Normalise keys before merging.

```suggestion
      # Merge .percy.yml config options with snapshot options (snapshot options take priority)
      config_options = (@cli_config&.dig('snapshot') || {}).transform_keys(&:to_sym)
      merged_options = config_options.merge(options)
```

_Reviewer: stack:code-review_

**Review by rishigupta1599 — COMMENTED** [review]
Claude Code Review (automated) — 1 inline finding(s). Full report in the PR comment below. Verdict: Failed - see PR comment.

File: lib/percy.rb (hunk 1 above)
Comment (rishigupta1599) [reply lib/percy.rb]: Fixed in 3a32b11. Normalized the key types before the merge: config keys (string, from JSON parse) are symbolized to match the per-call `options` (symbol keys) and all downstream consumers (`responsive_snapshot_capture?`, `capture_responsive_dom` read symbol keys). `config_options.transform_keys(&:to_sym).merge(options)` now yields one key per option with per-call options winning, and `.to_json` serializes identically.

File: lib/percy.rb [hunk 2]
```diff
@@ -79,10 +79,19 @@ def self.snapshot(driver, name, options = {})
...
-      dom_snapshot = if responsive_snapshot_capture?(options)
-        capture_responsive_dom(driver, options, percy_dom_script: percy_dom_script)
+
+      # Merge .percy.yml config options with snapshot options (snapshot options take priority)
+      config_options = @cli_config&.dig('snapshot') || {}
+      # Config keys are strings (JSON parse); per-call options use symbols, as
+      # do downstream consumers (responsive_snapshot_capture?, capture_responsive_dom).
+      # Symbolize config keys so per-call options override matching config keys
+      # instead of producing duplicate logical keys (per-call options take priority).
+      merged_options = config_options.transform_keys(&:to_sym).merge(options)
```
Comment (rishigupta1599) [inline lib/percy.rb]: **[Medium] Merge/key-override fix has no test coverage**

The core behavior of this fix — symbolizing config keys so per-call options override matching config keys — is not asserted by any spec. Existing serialize specs pass `{}` and never combine `@cli_config['snapshot']` with overlapping per-call options.

**Suggestion:** Add a unit test around `Percy.snapshot` that sets `@cli_config = {'snapshot' => {'percyCSS' => 'a', 'minHeight' => 700}}`, calls with `percyCSS: 'b'`, and asserts the serialized options contain `:percyCSS => 'b'` (per-call override) and `:minHeight => 700` (config inherited), with no duplicate string keys.

_Reviewer: stack:code-review_

**Review by rishigupta1599 — COMMENTED** [review]
Claude Code Review (automated) — 1 inline finding(s). Full report in the PR comment below. Verdict: Passed.

File: spec/lib/percy/percy_spec.rb [hunk 3]
```diff
@@ -297,6 +297,45 @@
 
       expect(data).to eq('sync_data')
     end
+
+    it 'merges .percy.yml config with per-snapshot options (per-call wins)' do
```
Comment (rishigupta1599) [inline spec/lib/percy/percy_spec.rb:346]: **[Medium] Test gap: config-only path and POST body unverified**

This test covers "config sets two keys, per-call overrides one" but not the motivating case where options come solely from `.percy.yml` (per-call `options = {}`). It also does not assert the `POST /percy/snapshot` body. The POST is built from raw `options` (`post_options = options.reject {...}`), not `merged_options`, so config-only keys reach `PercyDOM.serialize` but are not re-sent on POST. That asymmetry matches the existing healthcheck pattern but is currently untested.

**Suggestion:** Add a test with empty per-call options and a populated config `snapshot` block, asserting config keys reach `serialize`; optionally add a `have_requested(:post, ...)` assertion documenting the POST-body contract for config-derived keys.

_Reviewer: stack:code-review_

File: lib/percy.rb (hunk 2 above)
Comment (rishigupta1599) [inline lib/percy.rb]: **[Low] Shallow symbolize + now-redundant downstream config fallbacks**

`transform_keys(&:to_sym)` is shallow, so nested config hashes (e.g. a future `readiness` object) keep string keys. Current consumers re-normalize, so this is safe today — a forward-compat note, not a bug. Separately, `responsive_snapshot_capture?` and the `minHeight` read in `capture_responsive_dom` still keep their own `@cli_config&.dig(...)` fallbacks, now redundant given this upstream merge (harmless but a maintainability hazard).

**Suggestion:** Note in the comment that only top-level keys are symbolized; optionally drop the redundant downstream `@cli_config` fallbacks in a follow-up with a test confirming the merged hash is the single source of truth.

_Reviewer: stack:code-review_

**Review by rishigupta1599 — COMMENTED** [review]
Claude Code Review (automated) — 2 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Review by pranavz28 — APPROVED** [review]
